kafka-4295: ConsoleConsumer does not delete the temporary group in zookeeper

huxihx opened the pull request:
Since consumer stop logic and zk node removal code are in separate threads, so when two threads execute in an interleaving manner, persistent node '/consumers/' might not be removed for those console consumer groups which do not specify "group.id". This will pollute Zookeeper with lots of inactive console consumer offset information.

huxi-2b committed:
kafka-4295: kafka-console-consumer.sh does not delete the temporary group in zookeeper
Author: huxi

Since consumer stop logic and zk node removal code are in separate threads, so when two threads execute in an interleaving manner, persistent node '/consumers/<consumer-group>' might not be removed for those console consumer groups which do not specify "group.id". This will pollute Zookeeper with lots of inactive console consumer offset information.

huxihx commented:
@mjsax seems the failure is not relevant to the commit, how should I handle this situation? Please advice. thanks.

guozhangwang commented:
@amethystic Could you re-open the PR so that a Jenkins build can be triggered again? cc @hachikuji @ijuma for reviews.

huxihx commented:
@hachikuji @ijuma  please review this PR. Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -104,7 +104,10 @@ object ConsoleConsumer extends Logging {
-
+        // if we generated a random group id (as none specified explicitly) then avoid polluting zookeeper with persistent group data, this is a hack
+        if (!conf.groupIdPassed && conf.options.has(conf.zkConnectOpt))
+          ZkUtils.maybeDeletePath(conf.options.valueOf(conf.zkConnectOpt), "/consumers/" + conf.consumerProps.get("group.id"))
```
I think we have a utility in `AdminUtils` for this.

huxihx replied:
@hachikuji Yes, although AdminUtils already offers 'deleteConsumerGroupInZK' and 'deleteConsumerGroupInfoForTopicInZK' to implement this, I notice that ConsoleConsumer originally employ this snippet of code to delete inactive console consumer group.
The key point here is we must add this kind of code in addShutdownHook otherwise this thread might not delete the unused zk nodes if the first auto-commit task did not get started.

hachikuji replied:
Makes sense. It looks like `Adminutils.deleteConsumerGroupInZK` is the one we want. Perhaps we could replace both usages? Actually if we've added this to the shutdown hook, do we still need it in the `finally`?

huxihx replied:
The finally block already contains this cleanup call:
`consumer.cleanup()`
`conf.formatter.close()`
`reportRecordCount()`
`// if we generated a random group id (as none specified explicitly) then avoid polluting zookeeper with persistent group data, this is a hack
      if (!conf.groupIdPassed)
        ZkUtils.maybeDeletePath(conf.options.valueOf(conf.zkConnectOpt), "/consumers/" + conf.consumerProps.get("group.id"))
      shutdownLatch.countDown()`

ZkUtils.maybeDeletePath is idempotent, which can be called many times without any negative impact.  I call it from the shutdown hook thread to make sure this thread will try to delete the zkNode after the main thread fails to do this.

As for the question whether we should use deleteConsumerGroupInZK to replace ZkUtils.maybeDeletePath, I recommend we use the original one since only adding two lines in the shutdown hook thread is enough to fix the bug, we do not have to do any other regression test to make sure the new deleteConsumerGroupInZK  behaves as expected. But if you insist, I will close this PR and commit a new one with AdminUtils.deleteConsumerGroupInZK.
@hachikuji , what do you think?

hachikuji replied:
As far as I can tell, both methods delegate to `ZkClient.deleteRecursive`, and they both handle `ZkNoNodeException`, so I'm not sure I see the concern about idempotence. One difference is that `maybeDeletePath` catches a `Throwable`, but I don't see a good reason why we need to preserve that, especially since we're shutting down. Maybe I'm missing something?

Another question I have is about the `shutdownLatch.await()` below. If we're hitting a shutdown path where the `maybeDeletePath` is not being executed, wouldn't that mean we end up blocking in `await()`? Can you clarify the specific case that you're trying to handle in this patch?

Also, you don't need to close the PR if you want to change something. Just push a new commit. Your commits will get squashed when we merge anyway.

huxihx replied:
The reason why zk node failed to be removed is not that 'maybeDeletePath' is not executed, but because delete(path) got failed in deleteRecursive where parent znode got failed to be deleted if there still existed some children znode before calling delete(path). So if ConsoleConsumer ran exactly long enough to let autocommit thread creates the znode /consumers/<consumer_group>/offsets, everything is good. But if not, when shutting down the consumer, it commits the offset in the interval between deleting children znodes(namely, ids and owners) and deleting the parent znode (/consumers/***). Do I make myself clear?

Besides, I have already tested the patch, with such scenarios:
1. new consumer case: have to make sure there is no impact for the new consumer
2. old short-running consumer: successfully deleted
3. old long-running consumer: successfully deleted  
4. old short/long-running consumer with shut down by 'kill -9': Same behaviour as before, since if JVM process is shut down in this manner, event JVM shut down hook thread cannot be ensured to have a chance to run.
5. old consumer with groupid specified: have no impact since we only want to clear up those inactive groups without setting groupid

@hachikuji Does it make sense?

huxihx replied:
@hachikuji any feedbacks about this PR?

hachikuji replied:
Just one question:  do we need the additional call to `maybeDeletePath` in the `finally` block anymore? Adding it to the shutdown hook seems sufficient based on some testing. Is there any advantage to having it in both locations?

And just one note: it is actually possible to shutdown the console consumer before the shutdown hook gets registered, in which case you can still end up with the leftover node in Zk. You can try trigger this by shutting down the console consumer just after you see it register in zookeeper. This edge case is pretty tricky to address, so it's probably more important to get the main paths.

huxihx commented:
@joestein Could you help review this PR? Thanks.

huxihx commented:
@hachikuji Could you address the comments above? Thanks.

hachikuji commented:
@amethystic Apologies for the delay. I'll get to it today or tomorrow.

huxihx commented:
@hachikuji Any chances to address the comments these days? Please review the pull request. Thank you!

huxi-2b committed:
Merge branch 'kafka-4295_ConsoleConsumer_fail_to_remove_zknode_onexit' of https://github.com/amethystic/kafka into kafka-4295_ConsoleConsumer_fail_to_remove_zknode_onexit

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4295_ConsoleConsumer_fail_to_remove_zknode_onexit

huxi-2b committed:
kafka-ConsoleConsumer does not delete the temporary group in zookeeper
1. Remove same cleanup code from within the JVM shutdown hook code block
2. Refine ZkUtils.maybeDeletePath to capture ZkException if to-be-delete znode is not empty

huxi-2b committed:
kafka-4295: ConsoleConsumer does not delete the temporary group in zookeeper
remove useless imports in ZkUtils.scala

huxihx commented:
@hachikuji yes, you are right. As per your comments, I removed some unnecessary code for resource cleanup. ZkUtils.maybeDelete was also refined to capture the ZkException when the to-be-delete path is not empty. The fix only applies for the ConsoleConsumer shutdown via sending SIGINT and SIGTERM. For "kill -9", seems we cannot guarantee it works since application cannot capture it. Do all these make senses?

huxi-2b committed:
KAFKA-4428; Kafka does not exit if port is already bound

During Acceptor initialization, if "Address already in use" error is encountered,
the shutdown latch in each Processor is never counted down. Consequently,
the Kafka server hangs when `Processor.shutdown` is called.

Author: huxi <huxi@zhenrongbao.com>
Author: amethystic <huxi_2b@hotmail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2156 from amethystic/kafka-4428_Kafka_noexit_for_port_already_use

vahidhashemian committed:
MINOR: Sync up 'kafka-run-class.bat' with 'kafka-run-class.sh'

Some of the recent changes to `kafka-run-class.sh` have not been applied to `kafka-run-class.bat`.
These recent changes include setting proper streams or connect classpaths. So any streams or connect use case that leverages `kafka-run-class.bat` would fail with an error like
```
Error: Could not find or load main class org.apache.kafka.streams.???
```

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2238 from vahidhashemian/minor/sync_up_kafka-run-class.bat

eribeiro committed:
MINOR: Mx4jLoader always returns false even if mx4j is loaded & started

Mx4jLoader.scala should explicitly `return true` if the class is successfully loaded and started, otherwise it will return false even if the class is loaded.

Author: Edward Ribeiro <edward.ribeiro@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2295 from eribeiro/mx4jloader-bug

LoneRifle committed:
MINOR: Rephrase Javadoc summary for ConsumerRecord

The original Javadoc description for `ConsumerRecord` is slightly confusing in that it can be read in a way such that an object is a key value pair received from Kafka, but (only) consists of the metadata associated with the record. This PR makes it clearer that the metadata is _included_ with the record, and moves the comma so that the phrase "topic name and partition number" in the sentence is more closely associated with the phrase "from which the record is being received".

Author: LoneRifle <LoneRifle@users.noreply.github.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2290 from LoneRifle/patch-1

huxi-2b committed:
KAFKA-4351; MirrorMaker with new consumer should support comma-separated regex

This makes it consistent with MirrorMaker with the old consumer.

Author: huxi <huxi@zhenrongbao.com>
Author: amethystic <huxi_2b@hotmail.com>

Reviewers: Vahid Hashemian <vahidhashemian@us.ibm.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2072 from amethystic/kafka-4351_Regex_behavior_change_for_new_consumer

vahidhashemian committed:
KAFKA-4480: Report an error in 'kafka-configs' command if the config to be removed does not exist

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2218 from vahidhashemian/KAFKA-4480

huxihx commented:
@hachikuji please kindly take some time to review. Thanks.

himani1 committed:
MINOR: Scala code readability improvements

Author: Himani Arora <1himani.arora@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2297 from himani1/refactored_code

baluchicken committed:
KAFKA-4318; Migrate ProducerSendTest to the new consumer

Author: Balint Molnar <balintmolnar91@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2083 from baluchicken/KAFKA-4318

dguy committed:
KAFKA-4561; Ordering of operations in StreamThread.shutdownTasksAndState may void at-least-once guarantees

In `shutdownTasksAndState` and `suspendTasksAndState` we commit offsets BEFORE we flush any state. This is wrong as if an exception occurs during a flush, we may violate the at-least-once guarantees, that is we would have committed some offsets but NOT sent the processed data on to other Sinks.
Also during suspend and shutdown, we should try and complete all tasks even when exceptions occur. We should just keep track of the exception and rethrow it at the end if necessary. This helps with ensuring that StateStores etc are closed.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2281 from dguy/kafka-4561

Commit:
KAFKA-2434: Remove identical topic subscription constraint for roundrobin strategy in old consumer API

Author: Andrew Olson <aolson1@cerner.com>

Reviewers: Onur Karaman

Closes #145 from noslowerdna/KAFKA-2434

ijuma committed:
KAFKA-4528: Fix failure in ProducerTest.testAsyncSendCanCorrectlyFailWithTimeout

I was able to reproduce the failure in less than 10 runs before the change. With the change,
the test passed 70 times consecutively.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang

Closes #2298 from ijuma/kafka-4528-fix-test-async-send-timeout

hachikuji commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -89,12 +89,14 @@ object ZkUtils {
       zk.deleteRecursive(dir)
-      zk.close()
     } catch {
+      case _: ZkException => zk.deleteRecursive(dir)
```
Unfortunate that we cannot catch a more specific error here. Perhaps at least we can add a comment explaining the case that we're trying to handle?

huxihx replied:
already added some comments explaining why.

ewencp committed:
KAFKA-4404; Add javadocs to document core Connect types, especially that integer types are signed

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Konstantine Karantasis <konstantine@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2296 from ewencp/kafka-4404-document-connect-signed-integer-types

lindong28 committed:
KAFKA-4429; Consumer lag metric should be zero if FetchResponse is empty

Author: Dong Lin <lindong28@gmail.com>

Reviewers: Ewen Cheslack-Postava <me@ewencp.org>, Jason Gustafson <jason@confluent.io>

Closes #2155 from lindong28/KAFKA-4429

fpj committed:
KAFKA-4261: Provide debug option in vagrant-up.sh

Author: fpj <fpj@apache.org>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1981 from fpj/vagrant-debug-option

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4295_ConsoleConsumer_fail_to_remove_zknode_onexit

huxi-2b committed:
kafka-4295: ConsoleConsumer does not delete the temporary group in zookeeper

added some comments to explain why we capture ZkException when invoking deleteRecursive(dir) in maybeDeletePath

huxihx commented:
@hachikuji Thanks for your comments. Please kindly review it again.

huxi-2b committed:
kafka-4295: ConsoleConsumer does not delete the temporary group in zookeeper

added a unit test case to ensure no zknode left after consoleconsumer got exited.

huxihx commented:
@hachikuji added a unit test case to test that no zknode left after console consumer got exited. Please take some time to have a review on that. Many thanks.

ijuma commented:
What is the use case for using the console consumer with the ZK-based consumer? Generally, we've restricted ourselves to fixing critical bugs in the old clients and we are in the process of deprecating the old consumers via KIP-109 (the old producers are already deprecated).

huxihx commented:
@ijuma I am not sure if we should fix bugs for ZK-based consumer anymore, so do you suggest I should close the PR?

ijuma commented:
@amethystic, looking at the change, it seems like part of it actually fixes an issue when using the new consumer (we should not try to delete a path from the ZK in that case), so maybe we should try to include this.

ijuma commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -79,8 +79,9 @@ object ConsoleConsumer extends Logging {
 
       // if we generated a random group id (as none specified explicitly) then avoid polluting zookeeper with persistent group data, this is a hack
-      if (!conf.groupIdPassed)
+      if (!conf.groupIdPassed && conf.options.has(conf.zkConnectOpt)) {
```
`conf.useOldConsumer` is slightly better than checking `zkConnectOpt` directly.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -89,12 +89,14 @@ object ZkUtils {
       zk.deleteRecursive(dir)
-      zk.close()
     } catch {
+      case _: ZkException => zk.deleteRecursive(dir) // Occasionally, child node list is not empty before deleting parent ZkNode, give a second chance to try deleting.
```
I'm not sure about this. This is a generic method and I don't think this makes sense generally. I would actually remove `maybeDeletePath` as all callers should really be using `AdminUtils.deleteConsumerGroupInZK`.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -17,17 +17,41 @@
-import org.scalatest.junit.JUnitSuite
 
-class ConsoleConsumerTest extends JUnitSuite {
+class ConsoleConsumerTest extends KafkaServerTestHarness {
```
This change makes all the existing tests in this class _much_ slower. It would be better to create a separate test class (maybe `ConsoleConsumerIntegrationTest`) or to use mocks in this test.

ijuma reviewed (commented):
Thanks for the PR, left some comments.

huxi-2b committed:
kafka-4295: ConsoleConsumer does not delete the temporary group in
zookeeper

Addressed Ijuma's comments
1. Restored ZkUtils to trunk code
2. Restored ConsoleConsumerTest to trunk code
3. Restored ZkUtils.maybeDeletePath to trunk code
4. Replaced ZkUtils.maybeDeletePath with
AdminUtils.deleteConsumerGroupInZK

huxi-2b committed:
tweaked some imports to keep same as trunk

huxihx commented:
@ijuma Please review the PR again. Thanks.

huxi-2b committed:
change access level for ConsoleConsumer.deleteZkPathForConsumerGroup to private

huxi-2b committed:
changed access level for ConsoleConsumer.deleteZkPathForConsumerGroup to private[tools]x

huxi-2b committed:
removed useless import

huxihx commented:
@ijuma Please take time to review this PR. Thanks.

huxihx commented:
@ijuma @hachikuji  Please take some time to review the PR. Thanks.

huxihx commented:
@guozhangwang Do you know how to retest this fix since all the checks have failed although it is not known why it failed.

ijuma commented on core/src/test/scala/integration/kafka/tools/ConsoleConsumerIntegrationTest.scala:
```diff
@@ -0,0 +1,35 @@
+      assertTrue("Consumer group should be created.", zkUtils.getChildren(ZkUtils.ConsumersPath).head == groupID)
+    } finally {
+      consumer.stop()
+      ConsoleConsumer.deleteZkPathForConsumerGroup(conf.options.valueOf(conf.zkConnectOpt), conf.consumerProps.getProperty("group.id"))
```
It seems like the only thing this test is checking is that this call works. And since that is just calling `AdminUtils`, not sure if the benefit is worth it given the cost of starting up Kafka and ZK.

huxihx replied:
Agreed, but is there any way we could use to simulate sending terminate signals? This issue is caused by the fact `ZkUtils.maybeDeletePath` failed to delete the whole directory when ConsoleConsumer exits after receiving a INT or TERM signal , so what we can do is to test if the replaced method `deleteConsumerGroupInZK` works as expected. Any good idea to test this fix?

ijuma reviewed (commented):
Do I understand correctly that this PR now does the following:

1. Fix issue where we tried to delete a path in ZK for the new consumer.
2. Use `AdminUtils.deleteConsumerGroupInZK` instead of `ZKUtils.maybeDeletePath`

And the issue described in the JIRA remains since it's an edge case for the old consumer (that we intend to deprecate and remove)?

huxihx commented:
For your questions:
1: No, new consumer does not use ZK anymore. The fix is only for old consumer. Old consumer without group id specified leaves a lot of "garage" under /consumers due to the fact that `ZKUtils.deletePath` failed to delete them.
2: Yes. Tests showed that `AdminUtils.deleteConsumerGroupInZK` was able to clean up all the unused zk paths.

As what I said above, I am not sure if you guys still want to check in any code for fixing old consumer problems, especially when community is planning to remove it recently. It's up to you. If you don't, I am free to close the PR.
Does it make any sense?

huxihx commented:
@ijuma What's the status for this PR? Do I address all your comments already?

ijuma commented:
@huxihx sorry for the delay. I was going to look at this PR, but it seems like it includes other changes now. Maybe you could simply revive the part where we fix the new consumer not to call ZK unnecessarily?
